### PR TITLE
Move tiff JPEG compression outside the lock and thread it

### PR DESCRIPTION
**Pull request by jcupitt**
At the moment, libvips uses libtiff to write JPEG compressed tiles. Since compression is inside libtiff, it has to be single-threaded. 

This PR moves JPEG compression into libvips and uses a threadpool to write JPEG and JP2K compressed tiles in parallel.

git master:

```
$ time vips copy CMU-1.svs[rgb] x.tif[tile,compression=jpeg,pyramid]
real	0m11.440s
user	1m10.627s
sys	0m4.162s
```

This PR:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
real	0m4.972s
user	1m3.305s
sys	0m7.089s
```

### CODE REVIEW

**Comment by euzada** [comment]
Thank you @jcupitt for starting this PR. If I understand correctly, it still running on single thread even after move it inside libvips?

I failed to compile this version on ubuntu. There may be some issues compiling jpeg2vips.

Looking forward to test it.

**Comment by jcupitt** [comment]
TODO:

- [ ] only do this for tiled TIFFs
- [ ] support all JPEG TIFF write options eg. subsample mode etc.
- [ ] error handling
- [ ] check it still works without `HAVE_JPEG`
- [ ] all tests should pass
- [ ] changelog note
- [ ] clang-format

**Comment by jcupitt** [comment]
> Thank you @jcupitt for starting this PR. If I understand correctly, it still running on single thread even after move it inside libvips?

It compresses tiles in parallel, so you should see a speedup. 

It's not as dramatic as tiled jpeg tiff, because libvips already ran tiff compress in a background thread, but it does help a bit, it's perhaps 25% faster.

**Comment by jcupitt** [comment]
I added threaded write (as well as compression by libvips). 

Git master libvips:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile]
real	0m5.715s
user	0m52.887s
sys	0m4.017s
```

This PR:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile]
real	0m4.326s
user	0m56.713s
sys	0m6.424s
```

So about 30% faster on this PC.

**Comment by jcupitt** [comment]
This is the largest test image I have to hand:

```
$ time vips copy C2023000589S1-0-1_743eee76-f53b-aca2-4da5-fa97a44d9933_125232.svs[rgb] x.tif[tile,compression=jpeg,pyramid,tile-width=256,tile-height=256] --vips-progress --vips-leak
vips temp-2: 211213 x 87581 pixels, 32 threads, 128 x 128 tiles, 640 lines in buffer
vips temp-2: done in 86.9s          
vips_threadset_free: peak of 64 threads
memory: high-water mark 1.44 GB

real	1m27.625s
user	12m44.819s
sys	0m47.753s
```

**Comment by jcupitt** [comment]
Ah jp2k compression sees more of a speedup. Before:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jp2k,tile,pyramid]
real	7m40.086s
user	9m17.533s
sys	9m3.294s
```

After:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jp2k,tile,pyramid]
real	2m25.948s
user	10m17.321s
sys	18m49.749s
```

**Comment by euzada** [comment]
> Ah jp2k compression sees more of a speedup. Before:
> 
> ```
> $ time vips copy CMU-1.svs[rgb] x.tif[compression=jp2k,tile,pyramid]
> real	7m40.086s
> user	9m17.533s
> sys	9m3.294s
> ```
> 
> After:
> 
> ```
> $ time vips copy CMU-1.svs[rgb] x.tif[compression=jp2k,tile,pyramid]
> real	2m25.948s
> user	10m17.321s
> sys	18m49.749s
> ```
That again a very big improvement. So 70 % faster in case of jp2k and 35% faster in jpeg. 

Silly question : what's the advantage of using jp2k over jpeg in tiff compression? The jpeg tiles are very small and in term of their sizes, they are probably smaller and faster to load.

**Comment by jcupitt** [comment]
jp2k doesn't really have any advantages imo, it's just there for compatibility with some deranged software that still uses it.

**Comment by jcupitt** [comment]
Ah I found another speedup: we can copy the `TIFFTAG_JPEGTABLES` tag and skip decompress/recompress on level copy.

Git master:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
real	0m11.408s
user	1m10.671s
sys	0m4.546s
```

This PR:

```
$ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
real	0m4.972s
user	1m3.305s
sys	0m7.089s
```

More than 2x faster now.

**Comment by euzada** [comment]
> Ah I found another speedup: we can copy the `TIFFTAG_JPEGTABLES` tag and skip decompress/recompress on level copy.
> 
> Git master:
> 
> ```
> $ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
> real	0m11.408s
> user	1m10.671s
> sys	0m4.546s
> ```
> 
> This PR:
> 
> ```
> $ time vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
> real	0m4.972s
> user	1m3.305s
> sys	0m7.089s
> ```
> 
> More than 2x faster now.

Is this TIFFTAG_JPEGTABLES only optimized for Q=70 the default value?  [Tifftag_jpegtables](https://www.awaresystems.be/imaging/tiff/tifftags/jpegtables.html)

**Comment by jcupitt** [comment]
> Is this `TIFFTAG_JPEGTABLES` only optimized for Q=70 the default value? [Tifftag_jpegtables](https://www.awaresystems.be/imaging/tiff/tifftags/jpegtables.html)

It's created for each Q value.

**Comment by jcupitt** [comment]
OK, all done, I think.

**Comment by euzada** [comment]
> OK, all done, I think.

That was very fast for >50% improvement. Now to test it, do I need to compile this branch in mxe for windows?

**Comment by jcupitt** [comment]
> do I need to compile this branch in mxe for windows?

No idea, probably! I think it should work.

**Comment by euzada** [comment]
> ```
> ./build.sh --ref tiffsave-threaded-jpeg
> ```

Thank you. I will test it and come back to you.

**Comment by kleisauke** [comment]
It looks like this PR would cause non-deterministic output. See for example this test script:
```bash
vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]

expected_sha256=$(sha256sum "x.tif" | awk '{ print $1 }')
for run in {1..5}; do
  vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
  echo "$expected_sha256 x.tif" | sha256sum --check
done
```

Git master:
```console
$ ./check-deterministic.sh
x.tif: OK
x.tif: OK
x.tif: OK
x.tif: OK
x.tif: OK
```

This PR:
```console
$ ./check-deterministic.sh
x.tif: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
x.tif: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
x.tif: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
x.tif: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
x.tif: FAILED
sha256sum: WARNING: 1 computed checksum did NOT match
```

**Comment by euzada** [comment]
> It looks like this PR would cause non-deterministic output. See for example this test script:
> 
> ```shell
> vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
> 
> expected_sha256=$(sha256sum "x.tif" | awk '{ print $1 }')
> for run in {1..5}; do
>   vips copy CMU-1.svs[rgb] x.tif[compression=jpeg,tile,pyramid]
>   echo "$expected_sha256 x.tif" | sha256sum --check
> done
> ```
> 
> Git master:
> 
> ```
> $ ./check-deterministic.sh
> x.tif: OK
> x.tif: OK
> x.tif: OK
> x.tif: OK
> x.tif: OK
> ```
> 
> This PR:
> 
> ```
> $ ./check-deterministic.sh
> x.tif: FAILED
> sha256sum: WARNING: 1 computed checksum did NOT match
> x.tif: FAILED
> sha256sum: WARNING: 1 computed checksum did NOT match
> x.tif: FAILED
> sha256sum: WARNING: 1 computed checksum did NOT match
> x.tif: FAILED
> sha256sum: WARNING: 1 computed checksum did NOT match
> x.tif: FAILED
> sha256sum: WARNING: 1 computed checksum did NOT match
> ```

Is this test will verify the output hash? The size of the output is changing from run to run.

File: libvips/foreign/vips2tiff.c [hunk 1]
```diff
@@ -629,6 +643,167 @@ wtiff_embed_imagedescription(Wtiff *wtiff, TIFF *tif)
...
+
+	// disable chroma subsample for high Q
+	if (wtiff->Q >= 90) {
+		int i;
+
+		for (i = 0; i < image->Bands; i++) {
+			cinfo->comp_info[i].h_samp_factor = 1;
+			cinfo->comp_info[i].v_samp_factor = 1;
+		}
+	}
```
Comment (kleisauke) [inline libvips/foreign/vips2tiff.c:696]: When I omit the JFIF APP0 marker, I get file sizes similar to those of the Git master.
```suggestion
	}

	// Avoid writing the JFIF APP0 marker.
	cinfo->write_JFIF_header = FALSE;
```

**Comment by jcupitt** [comment]
> It looks like this PR would cause non-deterministic output.

Certainly -- it's writing the tiles across a row in parallel, so the tile order is set by which thread finishes first.

We could buffer all the compressed tiles in the row, then sort and write them sequentially, but it would obviously raise memory use.

File: libvips/foreign/vips2tiff.c (hunk 1 above)
Comment (jcupitt) [reply libvips/foreign/vips2tiff.c:696]: Ah nice! I forgot about this.

**Comment by jcupitt** [comment]
Thinking about it, you're right, non-deterministic output is pretty bad. I'll add gather-sort-write. Hopefully the increased memory use isn't too bad.

**Comment by jcupitt** [comment]
It should be sorted now.

`vips2tiff.c` is getting pretty big :( but maybe we can move parts of it out it later.

**Review by kleisauke — APPROVED** [review]
LGTM!

**Comment by euzada** [comment]
> LGTM!

Test it and it works well on windows

**Comment by jcupitt** [comment]
Great! Thanks for testing @euzada

**Pull request merged**